libass/ass_render.c

```diff
@@ -2079,6 +2080,24 @@ static void calculate_rotation_params(ASS_Renderer *render_priv, ASS_DRect *bbox
 }
 
 
+static int quantize_blur(double radius, int32_t *shadow_mask)
+{
+    const double scale = TRANFORM_ACCURACY / (64 * BLUR_ACCURACY);
+    radius *= 2 / sqrt(log(256)) / scale;
+
+    int ord;
+    frexp((1 + radius) * (TRANFORM_ACCURACY / 2), &ord);
+    *shadow_mask = ((uint32_t) 1 << ord) - 1;
```

Comment:
Only for insane blur radius around 10^9. Is that even possible to trigger with some broken input?

Reply:
Right, I guess we do limit the blur radius… to a measly 100. We should probably raise that limit quite a bit; ideally, remove it at all. But I digress. As it stands, we do currently limit the blur radius to 100 during parsing, so I guess the only way it can become huge here is if `blur_scale` is huge. Which is possible only if the screen height is huge and `PlayResY` or (if known) the video storage height is tiny. Assuming `PlayResY: 1`, the screen height would need to be about 10^7 to reach the value you mentioned. Well… I guess that’s pretty unlikely, but we should probably still handle that without overflowing.

Reply:
Large `ord` mean that we have insane blur radius. `ass_synth_blur` will break much earlier than `shadow_mask` overflows. I think it's better to limit final (after all transformations) blur radius.

Reply:
@rcombs Do you have an opinion on whether we should clip `ord` here to block overflow in the shift expression or just rely on the fact other things will presumably break earlier?

Reply:
We shouldn't have known code paths that lead to UB. If the resulting behavior is well-understood, we should clip; otherwise, we should add sane limits that prevent this from happening (seems like they could be quite high) and `assert()` that it doesn't.

Reply:
If I remember correctly, clipping this `ord` never affects correctness but only performance. The larger it is, the more cache entries will be merged.